File: misc/Pseudo.cpp
#include <iostream>
#include <sstream>

#include "Core/MIPS/MIPSDebugInterface.h"
#include "MyDocument.hpp"
#include "Pseudo.hpp"

struct PseudoTableRow {
  const char *mnemonic;
  int (MyDocument::*funcPseudo)(MyInstruction *instr);
};

static const std::map<std::string, PseudoTableRow> mapTables = {
  { "addiu", { "addiu", &MyDocument::PseuDoAssign } },
  { "addu", { "addu", &MyDocument::PseuDoAssign } },
  { "subu", { "subu", &MyDocument::PseuDoAssign } },

  { "move", { "move", &MyDocument::PseuDoAssign } },
  { "andi", { "andi", &MyDocument::PseuDoAssign } },
  { "ori", { "ori", &MyDocument::PseuDoAssign } },
  { "or", { "or", &MyDocument::PseuDoAssign } },
  { "xor", { "xor", &MyDocument::PseuDoAssign } },
  { "sll", { "sll", &MyDocument::PseuDoAssign } },
  { "sltiu", { "sltiu", &MyDocument::PseuDoAssign } },
  { "slti", { "slti", &MyDocument::PseuDoAssign } },
  { "sltu", { "sltu", &MyDocument::PseuDoAssign } },
  { "slt", { "slt", &MyDocument::PseuDoAssign } },
  { "sra", { "sra", &MyDocument::PseuDoAssign } },
  { "srl", { "sra", &MyDocument::PseuDoAssign } },

  { "lui", { "lui", &MyDocument::PseuDoLoadUpper } },
  { "lw", { "lw", &MyDocument::PseuDoLoad } },
  { "lbu", { "lbu", &MyDocument::PseuDoLoad }},
  { "li", { "li", &MyDocument::PseuDoAssign } },

  { "sw", { "sw", &MyDocument::PseuDoStore } },
  { "sb", { "sb", &MyDocument::PseuDoStore } },
  { "sh", { "sh", &MyDocument::PseuDoStore } },

  { "nop", { "nop", &MyDocument::PseuDoNothing } },

  { "beq", { "beq", &MyDocument::PseuDoJump } },
  { "beql", { "beql", &MyDocument::PseuDoJump } },
  { "bne", { "bne", &MyDocument::PseuDoJump } },
  { "bnel", { "bnel", &MyDocument::PseuDoJump } },
  { "blez", { "blez", &MyDocument::PseuDoJump } },
  { "bgtz", { "bgtz", &MyDocument::PseuDoJump } },
  { "bltz", { "bltz", &MyDocument::PseuDoJump } },
  { "bgez", { "bgez", &MyDocument::PseuDoJump } },

  { "jr", { "jr", &MyDocument::PseuDoJump } },
  { "j", { "j", &MyDocument::PseuDoJump } },
  { "jal", { "jal", &MyDocument::PseuDoJump } },

  { "syscall", { "syscall", &MyDocument::PseudoSyscall }}
};

int MyDocument::PseuDoNothing(MyInstruction *instr) {
  std::cout << "\t//" << std::endl;
  return 0;
}

int MyDocument::PseuDoAssign(MyInstruction *instr) {
  std::string op = "";
  bool arg2_is_dec = false;
  bool arg1_signed = false;
  bool arg2_signed = false;

  if (instr->mnemonic_ == "addiu") { op = "+"; }
  else if (instr->mnemonic_ == "addu" || instr->mnemonic_ == "li") { op = "+"; }
  else if (instr->mnemonic_ == "subu") { op = "-"; }
  else if (instr->mnemonic_ == "andi") { op = "&"; }
  else if (instr->mnemonic_ == "ori") { op = "|"; }
  else if (instr->mnemonic_ == "or") { op = "|"; }
  else if (instr->mnemonic_ == "xor") { op = "^"; }
  else if (instr->mnemonic_ == "sll") { op = "<<"; arg2_is_dec = true; }
  else if (instr->mnemonic_ == "sltiu") { op = "<"; }
  else if (instr->mnemonic_ == "sltu") { op = "<"; }
  else if (instr->mnemonic_ == "slti") { op = "<"; arg1_signed = true; }
  else if (instr->mnemonic_ == "slt") { op = "<"; arg1_signed = true; arg2_signed = true; }
  else if (instr->mnemonic_ == "sra") { op = ">>"; arg1_signed = true; arg2_is_dec = true; }
  else if (instr->mnemonic_ == "srl") { op = ">>"; arg2_is_dec = true; }

  if (instr->arguments_.size() > 2 && op.empty()) { return -1; }
  if (instr->arguments_.size() > 3) { return -1; }

  std::cout << '\t' << instr->arguments_[0].Str() << " = ";
  if (arg1_signed) {
    std::cout << "(s32)";
  }
  std::cout << instr->arguments_[1].Str();

  if (instr->arguments_.size() > 2 && !instr->arguments_[2].IsZero()) {
    if (!op.empty()) {
      std::cout << " " << op;
    }

    std::cout << " ";
    if (arg2_signed) {
      std::cout << "(s32)";
    }
    std::cout << instr->arguments_[2].Str(arg2_is_dec);
  }

  //std::cout << "\t// " << instr->AsString();
  std::cout << std::endl;

  return 0;
}

int MyDocument::PseuDoLoadUpper(MyInstruction *instr) {
  std::stringstream ss;
  ss << instr->arguments_[0].Str() << " = " << instr->arguments_[1].Str();

  if (instr->arguments_[1].IsZero()) { ; }
  if (instr->arguments_[1].IsNumber()) { ss << "0000"; }
  else { ss << " << 16"; }

  std::cout << '\t' << ss.str() << std::endl;
  return 0;
}

int MyDocument::PseuDoLoad(MyInstruction *instr) {
  std::string suffix = instr->mnemonic_.substr(1);
  std::string sz = "", mask = "";

  if (suffix == "w") { sz = "u32"; }
  else if (suffix == "bu") { sz = "u8"; mask = " & 0xff"; }
  else return -1;

  std::cout << "\t" << instr->arguments_[0].Str() << " = ("<<sz<<")" << instr->arguments_[1].Str();
  if (!mask.empty()) std::cout << mask;
  std::cout << std::endl;

  return 0;
}

int MyDocument::PseuDoStore(MyInstruction *instr) {
  std::string sz = "";
  if (instr->mnemonic_.back() == 'b') { sz = "u8"; }
  if (instr->mnemonic_.back() == 'h') { sz = "u16"; }
  if (instr->mnemonic_.back() == 'w') { sz = "u32"; }
  std::cout << "\t("<<sz<<")" << instr->arguments_[1].Str() << " = " << instr->arguments_[0].Str() << std::endl;
  return 0;
}

// return  0: ok,
//        -1: unimplemented,
//         1: skip delay shot.
//
int MyDocument::DumpPseudo(u32 addr) {
  if (!instrManager_.InstrIsExists(addr)) return -1;
  MyInstruction *instr = instrManager_.FetchInstruction(addr);
  auto it = mapTables.find(instr->mnemonic_);
  if (it == mapTables.end()) {
    std::cout << "\nWARNING\tDumpPseudo\tUnimplemented mnemonic: " << instr->mnemonic_ << std::endl;
    std::cout << instr->AsString(true) << std::endl;
    return -1;
  }
  return (this->*it->second.funcPseudo)(instr);
}

int MyDocument::PseuDoJump(MyInstruction *instr)
{
  int jal_ra = instr->addr_ + 4;
  if (instr->info_.hasDelaySlot) {
		jal_ra += 4;
  }
  auto &arg0 = instr->arguments_[0];
  MyInstruction *next_instr = nullptr;
  std::string op = "";
  std::string op_z = "";
  std::string op_l = "";

  if (instr->info_.hasDelaySlot) {
    next_instr = instrManager_.GetInstruction(instr->addr_ + 4);
    if (!next_instr) {
      std::cout << "WARNING\tPseuDoJump\tDelay shot instruction out of range" << std::endl;
      return -1;
    }
  }

  if (instr->mnemonic_ == "beq") op = "==";
  else if (instr->mnemonic_ == "bne") op = "!=";

  else if (instr->mnemonic_ == "blez") op_z = "<= 0";
  else if (instr->mnemonic_ == "bgtz") op_z = "> 0";
  else if (instr->mnemonic_ == "bltz") op_z = "< 0";
  else if (instr->mnemonic_ == "bgez") op_z = ">= 0";

  else if (instr->mnemonic_ == "bnel") op_l = "!=";
  else if (instr->mnemonic_ == "beql") op_l = "==";

  if (instr->mnemonic_ == "jal") {
    if (next_instr) DumpPseudo(next_instr->addr_);

    if (arg0.IsZero()) return -1;

    std::cout << "\tv0 = " << arg0.Str() << "(...)";
		std::cout << "\t// ra = " << jal_ra << ";";
		std::cout << " goto -> " << arg0.value_ << ";";
		std::cout << std::endl;
  } else if (instr->mnemonic_ == "j") {
    if (next_instr) DumpPseudo(next_instr->addr_);

    std::cout << "\tgoto -> " << arg0.Str() << std::endl;
  } else if (instr->mnemonic_ == "jr") {
    if (next_instr) DumpPseudo(next_instr->addr_);

    if (arg0.type_ == ArgReg && arg0.reg_ == "ra") {
			std::cout << "\treturn v0";
			std::cout << "\t// goto -> ra;";
			std::cout << std::endl;
		} else {
			std::cout << "\tgoto -> " << arg0.Str() << std::endl;
		}
  } else if (!op.empty()) {
		if (next_instr) DumpPseudo(next_instr->addr_);

    std::cout << '\t' << "if (" << instr->arguments_[0].Str() << " " << op << " " << instr->arguments_[1].Str() << ") ";
    std::cout << "goto -> " << instr->arguments_[2].Str() << std::endl;

  } else if (!op_z.empty()) {
		if (next_instr) DumpPseudo(next_instr->addr_);

    std::cout << '\t' << "if ((s32)" << instr->arguments_[0].Str() << " " << op_z << ") ";
		std::cout << "goto -> " << instr->arguments_[1].Str() << std::endl;
  } else if (!op_l.empty()) {
    std::cout << '\t' << "if (" << instr->arguments_[0].Str() << " " << op_l << " " << instr->arguments_[1].Str() << ") {" << std::endl;
		if (next_instr) {
			std::cout << '\t';
			DumpPseudo(next_instr->addr_);
		}
		std::cout << "\t\t" << "goto -> " << instr->arguments_[2].Str() << std::endl;
		std::cout << '\t' << "}" << std::endl;
  } else {
    return -1;
  }
	return instr->info_.hasDelaySlot ? +1 : 0;
}

int MyDocument::PseudoSyscall(MyInstruction *instr) {
  MyHLEFunction *hlefun = GetFunc(instr->arguments_[0].Str());
  if (hlefun) {
    std::cout << '\t';
    if (hlefun->retmask.size() > 0) {
      std::cout << "v0 = (" << hlefun->retmask[0] << ")";
    }

    std::cout << instr->arguments_[0].Str() << "(";

    for (int arg_i = 0; arg_i < hlefun->argmask.size(); arg_i++) {
      if (arg_i > 0)
        std::cout << ", ";
      std::cout << hlefun->argmask[arg_i] << " "
                << currentDebugMIPS->GetRegName(0, 4 + arg_i);
    }

    for (int ret_i = 1; ret_i < hlefun->retmask.size(); ret_i++) {
      if (ret_i > 1)
        std::cout << ", ";
      std::cout << hlefun->retmask[ret_i]
                  << "& v" << ret_i;
    }

    std::cout << ")" << std::endl;
  } else {
		std::cout << "\t" << instr->arguments_[0].Str() << "(...)" << std::endl;
    return -1;
  }
  return 0;
}

File: internal/pseudo.go
package internal

import (
	"fmt"
	"strconv"

	"github.com/firodj/pspsora/internal/codegen"
)

type PseudoPrint func(instr *SoraInstruction, doc *SoraDocument) (string, int)

var mnemonicToPseudo map[string]PseudoPrint

func init() {
	mnemonicToPseudo = map[string]PseudoPrint{
		"nop":   PseudoNothing,
		"addiu": PseudoAssign,
		"addu":  PseudoAssign,
		"subu":  PseudoAssign,

		"move":  PseudoAssign,
		"andi":  PseudoAssign,
		"ori":   PseudoAssign,
		"or":    PseudoAssign,
		"xor":   PseudoAssign,
		"sll":   PseudoAssign,
		"sltiu": PseudoAssign,
		"slti":  PseudoAssign,
		"sltu":  PseudoAssign,
		"slt":   PseudoAssign,
		"sra":   PseudoAssign,
		"srl":   PseudoAssign,

		"li":  PseudoAssign,
		"lui": PseudoLoadUpper,
		"lw":  PseudoLoad,
		"lbu": PseudoLoad,

		"sw": PseudoStore,
		"sb": PseudoStore,
		"sh": PseudoStore,

		"beq":  PseudoCondJump,
		"beql": PseudoCondJump,
		"bne":  PseudoCondJump,
		"bnel": PseudoCondJump,
		"blez": PseudoCondJump,
		"bgtz": PseudoCondJump,
		"bltz": PseudoCondJump,
		"bgez": PseudoCondJump,

		"jr":  PseudoJump,
		"j":   PseudoJump,
		"jal": PseudoJump,

		"syscall": PseudoSyscall,
	}
}

// return
//   - 0: ok,
//   - -1: unimplemented,
//   - 1: skip delay shot.
func Code(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	if fn, ok := mnemonicToPseudo[instr.Mnemonic]; ok {
		return fn(instr, doc)
	}
	return "", -1
}

func PseudoNothing(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	return "// nop", 0
}

func PseudoAssign(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	op := ""
	//arg2_is_dec := false
	arg1_signed := false
	arg2_signed := false

	switch instr.Mnemonic {
	case "addiu", "addu", "li":
		op = "+"
	case "subu":
		op = "-"
	case "andi":
		op = "&"
	case "ori", "or":
		op = "|"
	case "xor":
		op = "^"
	case "sll":
		op = "<<"
		//arg2_is_dec = true
	case "sltiu", "sltu":
		op = "<"
	case "slti":
		op = "<"
		arg1_signed = true
	case "slt":
		op = "<"
		arg1_signed = true
		arg2_signed = true
	case "sra":
		op = ">>"
		arg1_signed = true
		//arg2_is_dec = true
	case "srl":
		op = ">>"
		//arg2_is_dec = true
	}

	if len(instr.Args) > 2 && op == "" {
		panic("invalid number of argument")
	}

	if len(instr.Args) > 3 {
		panic("invalid number of argument")
	}

	s := codegen.ASTAssign{}

	s_left := instr.Args[0].ToPseudo()

	s_right := instr.Args[1].ToPseudo()

	if arg1_signed {
		s_right = &codegen.ASTUnary{
			Op:   "s32",
			Expr: s_right,
		}
	}

	if len(instr.Args) > 2 && !instr.Args[2].IsZero() {
		// Binary
		if op == "" {
			panic("missing op")
		}

		s_right1 := codegen.ASTBinary{
			Op:   op,
			Left: s_right,
		}

		s_right2 := instr.Args[2].ToPseudo()

		if arg2_signed {
			s_right2 = &codegen.ASTUnary{
				Op:   "s32",
				Expr: s_right2,
			}
		}

		s_right1.Right = s_right2
		s_right = &s_right1
	}

	s.Left = s_left
	s.Right = s_right

	return "--- " + s.String(), 0
}

func PseudoLoadUpper(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	s := codegen.ASTAssign{}

	s_left := instr.Args[0].ToPseudo()

	var s_right codegen.ASTNode

	s_right1 := instr.Args[1].ToPseudo()

	s_right2 := &codegen.ASTNumber{
		Value: 16,
	}

	s_right = &codegen.ASTBinary{
		Op:    "<<",
		Left:  s_right1,
		Right: s_right2,
	}
	s.Right = s_right

	s.Left = s_left
	s.Right = s_right

	return "--- " + s.String(), 0
}

func PseudoLoad(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	s := codegen.ASTAssign{}

	s_left := instr.Args[0].ToPseudo()
	s_right := instr.Args[1].ToPseudo()

	suffix := instr.Mnemonic[1:]
	switch suffix {
	case "w":
		s_right.(*codegen.ASTPointer).Sz = "u32"
	case "bu":
		s_right.(*codegen.ASTPointer).Sz = "u8"
		//mask = " & 0xff"
	case "hu":
		s_right.(*codegen.ASTPointer).Sz = "u16"
	case "b":
		s_right.(*codegen.ASTPointer).Sz = "s8"
		//mask = " & 0xff"
	case "h":
		s_right.(*codegen.ASTPointer).Sz = "s16"
	default:
		panic("unknown suffix: " + suffix)
	}

	s.Left = s_left
	s.Right = s_right

	return "--- " + s.String(), 0
}

func PseudoStore(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	s := codegen.ASTAssign{}

	s_left := instr.Args[1].ToPseudo()
	s_right := instr.Args[0].ToPseudo()

	suffix := instr.Mnemonic[1:]
	switch suffix {
	case "b":
		s_left.(*codegen.ASTPointer).Sz = "u8"
	case "h":
		s_left.(*codegen.ASTPointer).Sz = "u16"
	case "w":
		s_left.(*codegen.ASTPointer).Sz = "u32"
	default:
		panic("unknown suffix: " + suffix)
	}

	s.Left = s_left
	s.Right = s_right

	return "--- " + s.String(), 0
}

var maskToType map[byte]string = map[byte]string{
	'v': "void",
	'x': "u32",
	'i': "s32",
	'f': "float",
	'X': "u64",
	'I': "s64",
	'F': "double",
	's': "const char*",
	'p': "(u32*)",
}

func PseudoSyscall(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	moduleIndex, funcIndex := instr.GetSyscallNumber()

	//ss := ""

	_, fun := doc.GetHLE(moduleIndex, funcIndex)
	if fun != nil {
		s_name := codegen.ASTSymbolRef{}
		s_name.Name = doc.GetHLEFuncName(moduleIndex, funcIndex)
		s := codegen.ASTCall{
			Expr: &s_name,
		}

		for arg_i := range fun.ArgMask {
			//if arg_i > 0 {
			//	ss += ", "
			//}
			a_name := codegen.ASTSymbolRef{}
			a_name.Name = doc.GetRegName(0, 4+arg_i)
			arg := codegen.ASTUnary{
				Op:   maskToType[fun.ArgMask[arg_i]],
				Expr: &a_name,
			}
			//ss += "(" + maskToType[fun.ArgMask[arg_i]] + ")" + doc.GetRegName(0, 4+arg_i)
			s.Args = append(s.Args, &arg)
		}

		if len(fun.RetMask) > 0 {
			sa := codegen.ASTAssign{}
			sa.Right = &s

			for ret_i := range fun.RetMask {
				if ret_i == 0 {
					s_left := codegen.ASTSymbolRef{}
					s_left.Name = "v0"
					sa.Left = &s_left
					continue
				}

				//if ret_i > 1 {
				//	ss += ", "
				//}
				//ss += "(*" + maskToType[fun.RetMask[ret_i]] + ")&v" + strconv.Itoa(ret_i)
				ret_a := codegen.ASTSymbolRef{}
				ret_a.Name = "v" + strconv.Itoa(ret_i)
				arg := codegen.ASTUnary{
					Op:   "*" + maskToType[fun.RetMask[ret_i]] + "&",
					Expr: &ret_a,
				}
				s.Args = append(s.Args, &arg)
			}

			//s_left := codegen.ASTSymbolRef{}
			//s_left.Name = "v0"

			//ss += "v0 = (" + maskToType[fun.RetMask[0]] + ")"
			//s.Left = &s_left
		}

		//ss += +"("

		//ss += ")"
		return "--- " + s.String(), 0
	}

	s_name := codegen.ASTSymbolRef{}
	s_name.Name = fmt.Sprintf("m0x%02x::f0x%03x()", moduleIndex, funcIndex)
	s := codegen.ASTCall{
		Expr: &s_name,
	}

	return "--- " + s.String(), 0
}

func PseudoJump(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	jal_ra := instr.Address + 4
	if instr.Info.HasDelaySlot {
		jal_ra += 4
	}

	arg0 := instr.Args[0]
	ss := ""
	ss_next := ""

	if instr.Info.HasDelaySlot {
		next_instr := doc.InstrManager.Get(instr.Address + 4)
		if next_instr == nil {
			fmt.Printf("WARNING:\tPseudoJump delay shot instruction out of range\n")
			return "", -1
		}
		ss_next, _ = Code(next_instr, doc)
	}

	switch instr.Mnemonic {
	case "jal":
		if ss_next != "" {
			ss += ss_next + "\n"
		}

		if arg0.IsZero() {
			return "", -1
		}

		ss += "v0 = " + arg0.Str(false) + "(...);"
		ss += fmt.Sprintf("\t/* { ra = 0x%08x; ", jal_ra)
		ss += fmt.Sprintf("goto %s; } */", arg0.CodeLabel(doc))

	case "j":
		if ss_next != "" {
			ss += ss_next + "\n"
		}
		ss += fmt.Sprintf("goto %s;", arg0.CodeLabel(doc))
	case "jr":
		if ss_next != "" {
			ss += ss_next + "\n"
		}
		if arg0.Type == ArgReg && arg0.Reg == "ra" {
			ss += "return v0;"
			ss += "\t/* { goto -> ra; } */"
		} else {
			ss += fmt.Sprintf("goto %s;", arg0.Str(false))
		}
	default:
		panic("unknown jump")
	}

	if instr.Info.HasDelaySlot {
		return ss, 1
	}
	return ss, 0
}

func PseudoCondJump(instr *SoraInstruction, doc *SoraDocument) (string, int) {
	j_else := instr.Address + 4
	then_taken := false
	else_taken := false

	arg0 := instr.Args[0]

	op := ""
	op_l := ""
	ss := ""
	ss_next := ""

	var arg1 *SoraArgument = nil
	if len(instr.Args) >= 2 {
		arg1 = instr.Args[1]
	}

	if instr.Info.HasDelaySlot {
		next_instr := doc.InstrManager.Get(instr.Address + 4)
		if next_instr == nil {
			fmt.Printf("WARNING:\tPseudoCondJump delay shot instruction out of range\n")
			return "", -1
		}
		j_else += 4
		ss_next, _ = Code(next_instr, doc)
	} else {
		panic("unimplmented without delayslot")
	}

	theBB := doc.BBManager.Get(instr.Address)
	if theBB == nil {
		fmt.Printf("ERROR:\tunable to get BB for: 0x%08x", instr.Address)
		return "", -1
	}
	xref_tos := doc.BBManager.GetExitRefs(theBB.Address)
	for _, xref := range xref_tos {
		if xref.To == j_else {
			else_taken = true
		} else if !xref.IsAdjacent {
			then_taken = true
		}
	}

	switch instr.Mnemonic {
	case "beq":
		op = "=="
	case "bne":
		op = "!="
	case "blez":
		arg1 = NewSoraArgument("0", nil)
		op = "<="
	case "bgtz":
		arg1 = NewSoraArgument("0", nil)
		op = ">"
	case "bltz":
		arg1 = NewSoraArgument("0", nil)
		op = "<"
	case "bgez":
		arg1 = NewSoraArgument("0", nil)
		op = ">="
	case "bnel":
		op_l = "!="
	case "beql":
		op_l = "=="
	default:
		panic("unknown conditional jump")
	}

	argElse := NewSoraArgument(fmt.Sprintf("->$%08x", j_else), doc.SymMap.GetLabelName)

	if op != "" {
		ss += "zf = " + arg0.Str(false) + " " + op + " " + arg1.Str(false) + ";\n"
		ss += ss_next + ";\n"

		if then_taken && else_taken {
			ss += "if (zf) { goto " + instr.Args[2].CodeLabel(doc) + "; }"
		} else if then_taken {
			ss += "assert(zf); goto " + instr.Args[2].CodeLabel(doc) + ";"
		} else {
			ss += "assert(!zf); // goto " + argElse.CodeLabel(doc) + ";"
		}
	} else if op_l != "" {
		ss += "zf = " + arg0.Str(false) + " " + op_l + " " + arg1.Str(false) + ";\n"

		if then_taken && else_taken {
			ss += "if (!zf) { goto " + argElse.CodeLabel(doc) + "; }\n"
			ss += ss_next + "\n"
			ss += "goto " + instr.Args[2].CodeLabel(doc) + ";"
		} else if then_taken {
			ss += "assert(zf);\n"
			ss += ss_next + "\n"
			ss += "goto " + instr.Args[2].CodeLabel(doc) + ";"
		} else {
			ss += "assert(!zf); // goto " + argElse.CodeLabel(doc) + ";"
		}
	} else {
		return "", -1
	}

	if instr.Info.HasDelaySlot {
		return ss, 1
	}

	return ss, 0
}
